webhooks/virtualmachine/mutation/virtualmachine_mutator_intg_test.go

```diff
@@ -106,5 +106,60 @@ func intgTestsMutating() {
 
 		})
 
+		Context("SetDefaultPowerState", func() {
+			When("Creating VirtualMachine", func() {
+				When("When VM PowerState is empty", func() {
+					BeforeEach(func() {
+						vm.Spec.PowerState = ""
+					})
+					It("Should set PowerState to PoweredOn", func() {
+						Expect(ctx.Client.Create(ctx, vm)).To(Succeed())
+						modified := &vmopv1.VirtualMachine{}
+						Expect(ctx.Client.Get(ctx, client.ObjectKeyFromObject(vm), modified)).Should(Succeed())
+						Expect(modified.Spec.PowerState).Should(Equal(vmopv1.VirtualMachinePoweredOn))
+					})
+				})
+				When("When VM PowerState is not empty", func() {
+					BeforeEach(func() {
+						vm.Spec.PowerState = vmopv1.VirtualMachinePoweredOff
+					})
+					It("Should not mutate PowerState", func() {
+						Expect(ctx.Client.Create(ctx, vm)).To(Succeed())
+						modified := &vmopv1.VirtualMachine{}
+						Expect(ctx.Client.Get(ctx, client.ObjectKeyFromObject(vm), modified)).Should(Succeed())
+						Expect(modified.Spec.PowerState).Should(Equal(vmopv1.VirtualMachinePoweredOff))
+					})
+				})
+			})
+
+			When("Updating VirtualMachine", func() {
+				BeforeEach(func() {
+					vm.Spec.PowerState = vmopv1.VirtualMachinePoweredOff
+					Expect(ctx.Client.Create(ctx, vm)).To(Succeed())
+				})
+				When("When VM PowerState is empty", func() {
+					It("Should not mutate PowerState", func() {
+						modified := &vmopv1.VirtualMachine{}
+						Expect(ctx.Client.Get(ctx, client.ObjectKeyFromObject(vm), modified)).Should(Succeed())
+						modified.Spec.PowerState = ""
+						Expect(ctx.Client.Update(ctx, modified)).To(Succeed())
+						Expect(ctx.Client.Get(ctx, client.ObjectKeyFromObject(vm), modified)).Should(Succeed())
+						Expect(modified.Spec.PowerState).Should(BeEmpty())
```

Comment:
if it's not allowed to mutate powerstate when vm powerstate is empty, shouldn't this line to `Equal(vmopv1.VirtualMachinePoweredOff)`?

Reply:
Good catch! I'll add a code comment here. Basically there's no better way to validate this. This is specifically to validate that on update the value is not _automatically_ set to `poweredOn` if it is empty. Technically the original state is not possible, but I forced it to validate the update logic.